zkvm/src/predicate.rs

```diff
@@ -181,75 +165,46 @@ impl Into<CompressedRistretto> for Predicate {
     }
 }
 
-#[cfg(test)]
-mod tests {
-    use super::*;
-    use bulletproofs::PedersenGens;
-
-    fn bytecode(prog: &Program) -> Vec<u8> {
-        let mut prog_vec = Vec::new();
-        prog.encode(&mut prog_vec);
-        prog_vec
-    }
-
-    #[test]
-    fn valid_program_commitment() {
-        let prog = Program::build(|p| p.drop());
-        let pred = Predicate::unblinded_program(prog.clone());
-        let op = pred.prove_program_predicate(&bytecode(&prog), &[]);
-        assert!(op.verify().is_ok());
-    }
-
-    #[test]
-    fn invalid_program_commitment() {
-        let prog = Program::build(|p| p.drop());
-        let prog2 = Program::build(|p| p.dup(1));
-        let pred = Predicate::unblinded_program(prog);
-        let op = pred.prove_program_predicate(&bytecode(&prog2), &[]);
-        assert!(op.verify().is_err());
-    }
-
-    #[test]
-    fn valid_disjunction1() {
-        let gens = PedersenGens::default();
-
-        // dummy predicates
-        let preds = vec![
-            Predicate::Opaque(gens.B.compress()),
-            Predicate::Opaque(gens.B_blinding.compress()),
-        ];
-
-        let pred = Predicate::disjunction(preds.clone()).unwrap();
-        let op = pred.prove_disjunction(&preds);
-        assert!(op.verify().is_ok());
-    }
-
-    #[test]
-    fn valid_disjunction2() {
-        let gens = PedersenGens::default();
-
-        // dummy predicates
-        let preds = vec![
-            Predicate::Opaque(gens.B.compress()),
-            Predicate::Opaque(gens.B_blinding.compress()),
-        ];
-
-        let pred = Predicate::disjunction(preds.clone()).unwrap();
-        let op = pred.prove_disjunction(&preds);
-        assert!(op.verify().is_ok());
-    }
-    #[test]
-    fn invalid_disjunction1() {
-        let gens = PedersenGens::default();
-
-        // dummy predicates
-        let preds = vec![
-            Predicate::Opaque(gens.B_blinding.compress()),
-            Predicate::Opaque(gens.B.compress()),
-        ];
-
-        let pred = Predicate::Opaque(gens.B.compress());
-        let op = pred.prove_disjunction(&preds);
-        assert!(op.verify().is_err());
+impl CallProof {
+    pub fn serialized_length(&self) -> usize {
+        // VerificationKey is a 32-byte array
+        // MerkleNeighbor is a 32-byte array
+        32 + 4 + self.neighbors.len() * 32
+    }
+
+    pub fn parse(data: &[u8]) -> Result<Self, VMError> {
+        SliceReader::parse(data, |r| {
+            let mut call_proof = CallProof::default();
+            call_proof.signing_key = VerificationKey(r.read_point()?);
+            call_proof.positions = r.read_u32()?;
+            let mut neighbors = vec![];
+            while r.len() > 0 {
+                neighbors.push(r.read_u8x32()?);
+            }
+            for i in 0..neighbors.len() {
+                if call_proof.positions & (1 << 31 - i) == 0 {
```

Comment:
would be great to specify the endianness and semantics of the `callproof.positions`. This code seems to use the highest bit for the lowest node. Maybe reverse this? Then, the highest unused bits can be zeroes, so the overall number is ≈proportional to the proof size?